map_model/src/make/traffic_signals.rs

```diff
@@ -135,6 +140,7 @@ fn half_signal(map: &Map, i: IntersectionID) -> Option<ControlTrafficSignal> {
     for (id, movement) in &ts.movements {
         if id.crosswalk {
             ped_stage.edit_movement(movement, TurnPriority::Protected);
+            ped_stage.check_minimum_crosswalk_time(movement);
```

Comment:
There might be a few more places in this code in the future that construct Stages. How about we apply this transformation to everything generated? At the end of `get_possible_policies`, we can do something like:

```
    for (_, signal) in &mut results {
        for stage in &mut signal.stages {
            let crosswalks: Vec<MovementID> = stage.protected_movements.iter().filter(|id| id.crosswalk).cloned().collect();       
            for id in crosswalks {
                stage.check_minimum_crosswalk_time(&signal.movements[&id]);
            }
        }
    }
```

Reply:
So you would remove the `check_minimum_crosswalk_time` inside the stage constructors and just have a final check in get_possible_policies?
That's a pretty good idea and would make it easier to maintain/adding new stages.
I'll implement that now.
Thanks!

Reply:
The only way to achieve this is moving the `validate` function outside of the individual stage creation into `get_possible_policies` because of the `min_crossing_time` check.
Is this something we still want to do?

Reply:
Separated out the signal validation logic into a separate function, which gets called at the end of `get_possible_policies`.
Could do the same with the rest of the validation logic as well?

Reply:
Ah, I see the problem. If you call `validate_stage_timings` inside `validate`, it'll fail because `enforce_minimum_crosswalk_time` hasn't been called yet.

While importing, I think we want to first fix the timings, then validate. In the UI, the minimum time should already be enforced, so we just need to validate.

How about:
1) Inlining `validate_stage_timings` in `validate`
2) Changing all of the functions that generate signals to not perform validation. They still need to return `Option` because they might fail out-right for having the wrong number of roads, but we can consolidate the `.enforce_minimum()` then `.validate()` checks inside `get_policies`?

I didn't explain this very clearly, I'll check in a refactor and reply again

Reply:
https://github.com/dabreegster/abstreet/blob/2ddd9a62ffd2b426d142e6c98833c08afafe7b15/map_model/src/make/traffic_signals.rs#L64

I think it should be simpler to call `enforce_min` here, then add logic to `validate`

Reply:
Fixed, I think?